lib/core/src/Cardano/Pool/DB/Sqlite.hs

```diff
@@ -153,16 +156,49 @@ defaultFilePath = (</> "stake-pools.sqlite")
 -- library.
 withDBLayer
     :: Tracer IO PoolDbLog
+    -- ^ Logging object.
+    -> Maybe FilePath
+    -- ^ Database file location, or 'Nothing' for in-memory database.
+    -> TimeInterpreter IO
+    -- ^ The time interpreter object.
+    -> (DBLayer IO -> IO a)
+    -- ^ Action to run.
+    -> IO a
+withDBLayer = withDecoratedDBLayer undecoratedDB
+
+-- | A decorator for the database layer, useful for instrumenting or monitoring
+--   calls to database operations.
+newtype DBDecorator a =
+    DBDecorator { decorateDBLayer :: DBLayer a -> DBLayer a }
+
+-- | The identity decorator.
+--
+-- Equivalent to an undecorated database.
+--
+undecoratedDB :: DBDecorator a
+undecoratedDB = DBDecorator id
+
+-- | Runs an action with a connection to the SQLite database.
+--
+-- This function has the same behaviour as 'withDBLayer', but provides a way
+-- to decorate the created 'DBLayer' object with a 'DBDecorator', useful for
+-- instrumenting or monitoring calls to database operations.
+--
+withDecoratedDBLayer
+    :: DBDecorator IO
+       -- ^ The database decorator.
+    -> Tracer IO PoolDbLog
        -- ^ Logging object
     -> Maybe FilePath
        -- ^ Database file location, or Nothing for in-memory database
     -> TimeInterpreter IO
+       -- ^ The time interpreter object.
     -> (DBLayer IO -> IO a)
        -- ^ Action to run.
     -> IO a
-withDBLayer trace fp timeInterpreter action = do
+withDecoratedDBLayer dbDecorator trace fp timeInterpreter action = do
```

Comment:
We already have logs as _values_ in the database tracer, so I am not sure to understand the need for an additional decorator?

Reply:
@KtorZ wrote:
> We already have logs as _values_ in the database tracer, so I am not sure to understand the need for an additional decorator?

What mechanism do you propose for the integration test to get the values it needs?

Specifically, in the case of the `STAKE_POOLS_GARBAGE_COLLECTION_01` test:

* the test needs to examine all historical garbage collection events to check that there was indeed one garbage collection per epoch.
* by the time the test starts, several epochs might already have elapsed.
* if we only start recording events at the start of the test, then we won't be able to access earlier garbage collection events.
* the pool DB decorator is able to solve this problem as it starts recording garbage collection events from the very beginning of time.
* the pool DB decorator doesn't record anything else apart from garbage collection events, so the memory footprint should be minimal.

Reply:
We have used this approach with the logger for various other tests: 

- The logger can be instrumented to log message in a `TVar`
- That `TVar` can then be read at any time, and logs filtered from it to see for particular events. 

 > the pool DB decorator is able to solve this problem

Yes, but it adds yet-another-layer-of-complexity, for something for which we already had a working solution. It's not a huge addition and the impact on the normal production setup is quite small yet I am usually quite reluctant to add bits of code that are only used in testing.

Reply:
> We have used this approach with the logger for various other tests:
> The logger can be instrumented to log message in a TVar
> That TVar can then be read at any time, and logs filtered from it to see for particular events.

This sounds like a useful technique.

I can see there are some property tests that use this technique. (For example: `lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs`). Though I can't find any integration tests that use the same method.

For the purposes of `STAKE_POOLS_GARBAGE_COLLECTION_01`, it's important that we can inspect all the garbage collection events from the start of time, and not just those that occur during the duration of that one test case.

> Yes, but it adds yet-another-layer-of-complexity, for something for which we already had a working solution.

Sadly, no-one suggested this solution when we originally discussed this issue ([see original discussion](https://input-output-rnd.slack.com/archives/GBT05825V/p1600328272058800)), otherwise I would have been eager to try it out, especially if it would have meant that we didn't need to invent another abstraction.

Reply:
> We have used this approach with the logger for various other tests

I think:
- we _have_ used `traceInTVarIO` for testing small components in isolation.
- the problem here is to test a small component, _but_ in the context of the full integration tests. We'd need to setup `traceInTVarIO` globally in the cluster setup.

Although, I guess (🤔) a global `traceInTVarIO` + filter, would work _as well as_ the DB decorator… making it much better since we then don't need the "yet-another-layer-of-complexity".

Reply:
> Sadly, no-one suggested this solution when we originally discussed this issue

To me, doing stuff with `traceInTVarIO` or to inventing a DB Decorator seem both like implementation details of your first proposed solution:

> 1.  Make garbage collection happen once per epoch without fail (even when there are no new certificates), and then have our integration tests scan the logs for garbage collection messages. (Though we'd really want to avoid accumulating logs in memory.)